File: src/Rook/Serialization/RhinoSerializer.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using Rhino;
using Rhino.DocObjects;
using Rhino.Geometry;

namespace Rook.Serialization
{
    /// <summary>
    /// Serializes Rhino objects to JSON-friendly dictionaries.
    /// </summary>
    public static class RhinoSerializer
    {
        /// <summary>
        /// Serializes a RhinoObject to a dictionary.
        /// </summary>
        public static Dictionary<string, object?> SerializeObject(RhinoObject obj)
        {
            if (obj == null)
            {
                return new Dictionary<string, object?> { ["error"] = "Null object" };
            }

            var geometry = obj.Geometry;

            // Safely get layer name
            string layerPath = "Default";
            try
            {
                var doc = obj.Document;
                if (doc != null)
                {
                    var layerIndex = obj.Attributes.LayerIndex;
                    if (layerIndex >= 0 && layerIndex < doc.Layers.Count)
                    {
                        var layer = doc.Layers[layerIndex];
                        if (layer != null)
                        {
                            layerPath = layer.FullPath ?? layer.Name ?? "Default";
                        }
                    }
                }
            }
            catch
            {
                // Keep default "Default" layer name
            }

            var result = new Dictionary<string, object?>
            {
                ["id"] = obj.Id.ToString(),
                ["type"] = obj.ObjectType.ToString(),
                ["layer"] = layerPath,
                ["name"] = string.IsNullOrEmpty(obj.Attributes.Name) ? null : obj.Attributes.Name,
                ["visible"] = obj.Visible,
                ["bbox"] = geometry != null ? SerializeBoundingBox(geometry.GetBoundingBox(true)) : null
            };

            // Add color if not by layer
            if (obj.Attributes.ColorSource == ObjectColorSource.ColorFromObject)
            {
                result["color"] = SerializeColor(obj.Attributes.ObjectColor);
            }

            return result;
        }

        /// <summary>
        /// Serializes a RhinoObject with detailed geometry information.
        /// </summary>
        public static Dictionary<string, object?> SerializeObjectDetailed(RhinoObject obj)
        {
            var result = SerializeObject(obj);
            result["geometry"] = SerializeGeometry(obj.Geometry);
            result["attributes"] = SerializeAttributes(obj.Attributes);
            return result;
        }

        /// <summary>
        /// Serializes geometry details based on type.
        /// </summary>
        public static Dictionary<string, object?> SerializeGeometry(GeometryBase? geometry)
        {
            if (geometry == null)
            {
                return new Dictionary<string, object?> { ["error"] = "Null geometry" };
            }

            var result = new Dictionary<string, object?>
            {
                ["type"] = geometry.ObjectType.ToString()
            };

            switch (geometry)
            {
                case Rhino.Geometry.Point point:
                    result["location"] = SerializePoint(point.Location);
                    break;

                case PointCloud cloud:
                    result["pointCount"] = cloud.Count;
                    break;

                case LineCurve line:
                    result["start"] = SerializePoint(line.PointAtStart);
                    result["end"] = SerializePoint(line.PointAtEnd);
                    result["length"] = Math.Round(line.GetLength(), 4);
                    break;

                case PolylineCurve polyline:
                    result["pointCount"] = polyline.PointCount;
                    result["length"] = Math.Round(polyline.GetLength(), 4);
                    result["isClosed"] = polyline.IsClosed;
                    break;

                case ArcCurve arc:
                    result["center"] = SerializePoint(arc.Arc.Center);
                    result["radius"] = Math.Round(arc.Arc.Radius, 4);
                    result["angle"] = Math.Round(arc.Arc.AngleDegrees, 2);
                    result["length"] = Math.Round(arc.GetLength(), 4);
                    result["isClosed"] = arc.IsClosed;
                    result["degree"] = arc.Degree;
                    break;

                case NurbsCurve nurbsCurve:
                    result["length"] = Math.Round(nurbsCurve.GetLength(), 4);
                    result["isClosed"] = nurbsCurve.IsClosed;
                    result["degree"] = nurbsCurve.Degree;
                    result["pointCount"] = nurbsCurve.Points.Count;
                    result["domain"] = new[] { nurbsCurve.Domain.T0, nurbsCurve.Domain.T1 };
                    // Check if it's a circle
                    if (nurbsCurve.TryGetCircle(out var circle))
                    {
                        result["isCircle"] = true;
                        result["center"] = SerializePoint(circle.Center);
                        result["radius"] = Math.Round(circle.Radius, 4);
                    }
                    break;

                case Curve curve:
                    result["length"] = Math.Round(curve.GetLength(), 4);
                    result["isClosed"] = curve.IsClosed;
                    result["degree"] = curve.Degree;
                    result["domain"] = new[] { curve.Domain.T0, curve.Domain.T1 };
                    break;

                case Brep brep:
                    result["faceCount"] = brep.Faces.Count;
                    result["edgeCount"] = brep.Edges.Count;
                    result["vertexCount"] = brep.Vertices.Count;
                    result["isSolid"] = brep.IsSolid;
                    result["isManifold"] = brep.IsManifold;
                    if (brep.IsSolid)
                    {
                        var volume = brep.GetVolume();
                        if (!double.IsNaN(volume))
                            result["volume"] = Math.Round(volume, 4);
                    }
                    var area = brep.GetArea();
                    if (!double.IsNaN(area))
                        result["area"] = Math.Round(area, 4);
                    break;

                case Extrusion extrusion:
                    result["isSolid"] = extrusion.IsSolid;
                    result["capCount"] = extrusion.CapCount;
                    break;

                case Mesh mesh:
                    result["vertexCount"] = mesh.Vertices.Count;
                    result["faceCount"] = mesh.Faces.Count;
                    result["isClosed"] = mesh.IsClosed;
                    break;

                case Surface surface:
                    result["isClosed"] = surface.IsClosed(0) || surface.IsClosed(1);
                    result["domainU"] = new[] { surface.Domain(0).T0, surface.Domain(0).T1 };
                    result["domainV"] = new[] { surface.Domain(1).T0, surface.Domain(1).T1 };
                    break;
            }

            return result;
        }

        /// <summary>
        /// Serializes a Layer to a dictionary.
        /// </summary>
        public static Dictionary<string, object?> SerializeLayer(Layer layer)
        {
            return new Dictionary<string, object?>
            {
                ["id"] = layer.Id.ToString(),
                ["index"] = layer.Index,
                ["name"] = layer.Name,
                ["fullPath"] = layer.FullPath,
                ["color"] = SerializeColor(layer.Color),
                ["visible"] = layer.IsVisible,
                ["locked"] = layer.IsLocked,
                ["parentId"] = layer.ParentLayerId == Guid.Empty ? null : layer.ParentLayerId.ToString()
            };
        }

        /// <summary>
        /// Serializes object attributes.
        /// </summary>
        public static Dictionary<string, object?> SerializeAttributes(ObjectAttributes attrs)
        {
            var result = new Dictionary<string, object?>
            {
                ["name"] = attrs.Name,
                ["colorSource"] = attrs.ColorSource.ToString(),
                ["layerIndex"] = attrs.LayerIndex,
                ["materialIndex"] = attrs.MaterialIndex,
                ["linetype"] = attrs.LinetypeIndex
            };

            // Include user strings if any
            var userStrings = attrs.GetUserStrings();
            if (userStrings.Count > 0)
            {
                var dict = new Dictionary<string, string>();
                foreach (string key in userStrings.AllKeys)
                {
                    dict[key] = userStrings[key] ?? "";
                }
                result["userStrings"] = dict;
            }

            return result;
        }

        /// <summary>
        /// Serializes a Point3d to an array [x, y, z].
        /// </summary>
        public static double[] SerializePoint(Point3d point)
        {
            return new[] { Math.Round(point.X, 4), Math.Round(point.Y, 4), Math.Round(point.Z, 4) };
        }

        /// <summary>
        /// Serializes a BoundingBox to min/max arrays.
        /// </summary>
        public static Dictionary<string, double[]> SerializeBoundingBox(BoundingBox bbox)
        {
            return new Dictionary<string, double[]>
            {
                ["min"] = SerializePoint(bbox.Min),
                ["max"] = SerializePoint(bbox.Max)
            };
        }

        /// <summary>
        /// Serializes a Color to RGB dictionary.
        /// </summary>
        public static Dictionary<string, int> SerializeColor(Color color)
        {
            return new Dictionary<string, int>
            {
                ["r"] = color.R,
                ["g"] = color.G,
                ["b"] = color.B
            };
        }
    }
}

File: src/RookNative/Serialization/RhinoSerializer.cpp
// RhinoSerializer.cpp
//
// Snapshot → JSON conversion. Runs on worker threads (no Rhino SDK access).

#include "stdafx.h"
#include "Serialization/RhinoSerializer.h"

namespace Rook {
namespace Serializer {

// --- Shared enum-to-wire-string mappings ---

// Canonical mapping — promoted 2026-04-19 from static-in-GeometryHandler.cpp
// per Codex PR #64/#65 review. Four callers at promotion time:
// GeometryHandler::CaptureGeometryDetail, HandleDimLinear, HandleDimAligned,
// HandleDimRadius/HandleDimDiameter (PR-4). Keep strings stable — wire
// contract with agents.
std::string AnnotationTypeToString(ON::AnnotationType annotationType)
{
    switch (annotationType)
    {
    case ON::AnnotationType::Text:       return "Text";
    case ON::AnnotationType::Leader:     return "Leader";
    case ON::AnnotationType::Aligned:    return "AlignedDimension";
    case ON::AnnotationType::Angular:    return "AngularDimension";
    case ON::AnnotationType::Angular3pt: return "Angular3ptDimension";
    case ON::AnnotationType::Diameter:   return "DiameterDimension";
    case ON::AnnotationType::Radius:     return "RadialDimension";
    case ON::AnnotationType::Rotated:    return "LinearDimension";
    case ON::AnnotationType::Ordinate:   return "OrdinateDimension";
    case ON::AnnotationType::ArcLen:     return "ArcLengthDimension";
    case ON::AnnotationType::CenterMark: return "Centermark";
    default:                             return "Annotation";
    }
}

// --- Primitives ---

nlohmann::json SerializeColor(const ColorSnapshot& c)
{
    return { {"r", c.r}, {"g", c.g}, {"b", c.b} };
}

nlohmann::json SerializeBBox(const BBoxSnapshot& b)
{
    return {
        {"min", { b.min[0], b.min[1], b.min[2] }},
        {"max", { b.max[0], b.max[1], b.max[2] }}
    };
}

nlohmann::json SerializePoint3(const Point3& pt)
{
    return { pt.x, pt.y, pt.z };
}

// --- Object summary ---

nlohmann::json SerializeObject(const ObjectSnapshot& obj)
{
    nlohmann::json j;
    j["id"] = obj.id;
    j["type"] = obj.type;
    j["layer"] = obj.layer;
    j["name"] = obj.name.empty() ? nlohmann::json(nullptr) : nlohmann::json(obj.name);
    j["visible"] = obj.visible;
    j["bbox"] = SerializeBBox(obj.bbox);

    if (obj.color.has_value())
        j["color"] = SerializeColor(*obj.color);

    if (!obj.blockName.empty())
    {
        j["blockName"] = obj.blockName;
        j["blockDefinitionId"] = obj.blockDefinitionId;
    }

    return j;
}

// --- Geometry detail visitor (declared in RhinoSerializer.h) ---

nlohmann::json GeometryDetailVisitor::operator()(const BrepDetail& d) const
{
    nlohmann::json j;
    j["type"] = "Brep";
    j["faceCount"] = d.faceCount;
    j["edgeCount"] = d.edgeCount;
    j["vertexCount"] = d.vertexCount;
    j["isSolid"] = d.isSolid;
    j["isManifold"] = d.isManifold;
    if (d.volume.has_value()) j["volume"] = *d.volume;
    if (d.area.has_value()) j["area"] = *d.area;
    return j;
}

nlohmann::json GeometryDetailVisitor::operator()(const MeshDetail& d) const
{
    return {
        {"type", "Mesh"},
        {"vertexCount", d.vertexCount},
        {"faceCount", d.faceCount},
        {"isClosed", d.isClosed}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const LineCurveDetail& d) const
{
    return {
        {"type", "LineCurve"},
        {"start", SerializePoint3(d.start)},
        {"end", SerializePoint3(d.end)},
        {"length", d.length}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const PolylineCurveDetail& d) const
{
    return {
        {"type", "PolylineCurve"},
        {"pointCount", d.pointCount},
        {"length", d.length},
        {"isClosed", d.isClosed}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const ArcCurveDetail& d) const
{
    return {
        {"type", "ArcCurve"},
        {"center", SerializePoint3(d.center)},
        {"radius", d.radius},
        {"angle", d.angle},
        {"length", d.length},
        {"isClosed", d.isClosed},
        {"degree", d.degree}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const NurbsCurveDetail& d) const
{
    nlohmann::json j;
    j["type"] = "NurbsCurve";
    j["length"] = d.length;
    j["isClosed"] = d.isClosed;
    j["degree"] = d.degree;
    j["pointCount"] = d.pointCount;
    j["domain"] = { d.domain[0], d.domain[1] };

    if (d.isCircle)
    {
        j["isCircle"] = true;
        if (d.center.has_value())
            j["center"] = SerializePoint3(*d.center);
        if (d.radius.has_value())
            j["radius"] = *d.radius;
    }

    return j;
}

nlohmann::json GeometryDetailVisitor::operator()(const GenericCurveDetail& d) const
{
    return {
        {"type", "Curve"},
        {"length", d.length},
        {"isClosed", d.isClosed},
        {"degree", d.degree},
        {"domain", { d.domain[0], d.domain[1] }}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const SurfaceDetail& d) const
{
    return {
        {"type", "Surface"},
        {"isClosed", d.isClosed},
        {"domainU", { d.domainU[0], d.domainU[1] }},
        {"domainV", { d.domainV[0], d.domainV[1] }}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const ExtrusionDetail& d) const
{
    return {
        {"type", "Extrusion"},
        {"isSolid", d.isSolid},
        {"capCount", d.capCount}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const PointDetail& d) const
{
    return {
        {"type", "Point"},
        {"location", SerializePoint3(d.location)}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const PointCloudDetail& d) const
{
    return {
        {"type", "PointCloud"},
        {"pointCount", d.pointCount}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const SubDDetail& d) const
{
    return {
        {"type", "SubD"},
        {"vertexCount", d.vertexCount},
        {"edgeCount", d.edgeCount},
        {"faceCount", d.faceCount}
    };
}

nlohmann::json GeometryDetailVisitor::operator()(const AnnotationDetail& d) const
{
    return {
        {"type", "Annotation"},
        {"annotationType", d.annotationType},
        {"plainText", d.plainText},
        {"richText", d.richText},
        {"fontFamily", d.fontFamily},
        {"fontFace", d.fontFace},
        {"bold", d.bold},
        {"italic", d.italic},
        {"textHeight", d.textHeight}
    };
}

// --- Attributes ---

static nlohmann::json SerializeAttributes(const AttributeSnapshot& attrs)
{
    nlohmann::json j;
    j["name"] = attrs.name.empty() ? nlohmann::json(nullptr) : nlohmann::json(attrs.name);
    j["colorSource"] = attrs.colorSource;
    j["layerIndex"] = attrs.layerIndex;
    j["materialIndex"] = attrs.materialIndex;
    j["linetype"] = attrs.linetypeIndex;

    if (!attrs.userStrings.empty())
    {
        nlohmann::json us = nlohmann::json::object();
        for (const auto& [key, val] : attrs.userStrings)
            us[key] = val;
        j["userStrings"] = std::move(us);
    }

    return j;
}

// --- Detailed geometry ---

nlohmann::json SerializeGeometryDetailed(const GeometrySnapshot& snap)
{
    nlohmann::json j;

    // Base object info
    j["id"] = snap.id;
    j["type"] = snap.objectType;
    j["layer"] = snap.layer;
    j["name"] = snap.name.empty() ? nlohmann::json(nullptr) : nlohmann::json(snap.name);
    j["visible"] = snap.visible;
    j["bbox"] = SerializeBBox(snap.bbox);

    if (snap.color.has_value())
        j["color"] = SerializeColor(*snap.color);

    // Geometry details via variant visitor
    j["geometry"] = std::visit(GeometryDetailVisitor{}, snap.detail);

    // Attributes
    j["attributes"] = SerializeAttributes(snap.attributes);

    return j;
}

// --- Layer ---

nlohmann::json SerializeLayer(const LayerSnapshot& layer)
{
    nlohmann::json j;
    j["id"] = layer.id;
    j["index"] = layer.index;
    j["name"] = layer.name;
    j["fullPath"] = layer.fullPath;
    j["color"] = SerializeColor(layer.color);
    j["plotColor"] = SerializeColor(layer.plotColor);
    j["plotWeight"] = layer.plotWeight;
    j["linetype"] = layer.linetypeName;
    j["linetypeIndex"] = layer.linetypeIndex;
    j["material"] = layer.materialName.empty()
        ? nlohmann::json(nullptr)
        : nlohmann::json(layer.materialName);
    j["materialIndex"] = layer.materialIndex;
    j["visible"] = layer.visible;
    j["locked"] = layer.locked;
    j["expanded"] = layer.expanded;
    j["parentId"] = layer.parentId.empty()
        ? nlohmann::json(nullptr)
        : nlohmann::json(layer.parentId);
    j["objectCount"] = layer.objectCount;
    return j;
}

// --- Document ---

nlohmann::json SerializeDocument(const DocumentSnapshot& doc)
{
    return {
        {"documentSerialNumber", doc.documentSerialNumber},
        {"name", doc.name},
        {"path", doc.path},
        {"units", doc.units},
        {"tolerance", doc.tolerance},
        {"angleTolerance", doc.angleTolerance},
        {"objectCount", doc.objectCount},
        {"layerCount", doc.layerCount},
        {"activeLayer", doc.activeLayer},
        {"modified", doc.modified}
    };
}

} // namespace Serializer
} // namespace Rook
